File: Smile/src/main/java/smile/mds/MDS.java
/*******************************************************************************
 * Copyright (c) 2010 Haifeng Li
 *   
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *  
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
package smile.mds;

import smile.math.Math;
import smile.math.matrix.EigenValueDecomposition;

/**
 * Classical multidimensional scaling, also known as principal coordinates
 * analysis. Given a matrix of dissimilarities (e.g. pairwise distances), MDS
 * finds a set of points in low dimensional space that well-approximates the
 * dissimilarities in A. We are not restricted to using a Euclidean
 * distance metric. However, when Euclidean distances are used MDS is
 * equivalent to PCA.
 *
 * @see smile.projection.PCA
 * @see SammonMapping
 * 
 * @author Haifeng Li
 */
public class MDS {

    /**
     * Component scores.
     */
    private double[] eigenvalues;
    /**
     * Coordinate matrix.
     */
    private double[][] coordinates;
    /**
     * The proportion of variance contained in each principal component.
     */
    private double[] proportion;

    /**
     * Returns the component scores, ordered from largest to smallest.
     */
    public double[] getEigenValues() {
        return eigenvalues;
    }

    /**
     * Returns the proportion of variance contained in each eigenvectors,
     * ordered from largest to smallest.
     */
    public double[] getProportion() {
        return proportion;
    }

    /**
     * Returns the principal coordinates of projected data.
     */
    public double[][] getCoordinates() {
        return coordinates;
    }

    /**
     * Constructor. Learn the classical multidimensional scaling.
     * Map original data into 2-dimensional Euclidean space.
     * @param proximity the nonnegative proximity matrix of dissimilarities. The
     * diagonal should be zero and all other elements should be positive and
     * symmetric. For pairwise distances matrix, it should be just the plain
     * distance, not squared.
     */
    public MDS(double[][] proximity) {
        this(proximity, 2);
    }

    /**
     * Constructor. Learn the classical multidimensional scaling.
     * @param proximity the nonnegative proximity matrix of dissimilarities. The
     * diagonal should be zero and all other elements should be positive and
     * symmetric. For pairwise distances matrix, it should be just the plain
     * distance, not squared.
     * @param k the dimension of the projection.
     */
    public MDS(double[][] proximity, int k) {
        this(proximity, k, false);
    }

    /**
     * Constructor. Learn the classical multidimensional scaling.
     * @param proximity the nonnegative proximity matrix of dissimilarities. The
     * diagonal should be zero and all other elements should be positive and
     * symmetric. For pairwise distances matrix, it should be just the plain
     * distance, not squared.
     * @param k the dimension of the projection.
     * @param add true to estimate an appropriate constant to be added
     * to all the dissimilarities, apart from the self-dissimilarities, that
     * makes the learning matrix positive semi-definite. The other formulation of
     * the additive constant problem is as follows. If the proximity is
     * measured in an interval scale, where there is no natural origin, then there
     * is not a sympathy of the dissimilarities to the distances in the Euclidean
     * space used to represent the objects. In this case, we can estimate a constant c
     * such that proximity + c may be taken as ratio data, and also possibly
     * to minimize the dimensionality of the Euclidean space required for
     * representing the objects.
     */
    public MDS(double[][] proximity, int k, boolean add) {
        int m = proximity.length;
        int n = proximity[0].length;

        if (m != n) {
            throw new IllegalArgumentException("The proximity matrix is not square.");
        }

        if (k < 1 || k >= n) {
            throw new IllegalArgumentException("Invalid k = " + k);
        }

        double[][] A = new double[n][n];
        double[][] B = new double[n][n];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                A[i][j] = -0.5 * Math.sqr(proximity[i][j]);
                A[j][i] = A[i][j];
            }
        }

        double[] mean = Math.rowMean(A);
        double mu = Math.mean(mean);

        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= i; j++) {
                B[i][j] = A[i][j] - mean[i] - mean[j] + mu;
                B[j][i] = B[i][j];
            }
        }

        if (add) {
            double[][] Z = new double[2 * n][2 * n];
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    Z[i][n + j] = 2 * B[i][j];
                }
            }

            for (int i = 0; i < n; i++) {
                Z[n + i][i] = -1;
            }

            mean = Math.rowMean(proximity);
            mu = Math.mean(mean);
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    Z[n + i][n + j] = 2 * (proximity[i][j] - mean[i] - mean[j] + mu);
                }
            }

            EigenValueDecomposition eigen = Math.eigen(Z, false, true);
            double c = Math.max(eigen.getEigenValues());

            for (int i = 0; i < n; i++) {
                B[i][i] = 0.0;
                for (int j = 0; j < i; j++) {
                    B[i][j] = -0.5 * Math.sqr(proximity[i][j] + c);
                    B[j][i] = B[i][j];
                }
            }
        }

        EigenValueDecomposition eigen = Math.eigen(B, k);
        
        coordinates = new double[n][k];
        for (int j = 0; j < k; j++) {
            if (eigen.getEigenValues()[j] < 0) {
                throw new IllegalArgumentException(String.format("Some of the first %d eigenvalues are < 0.", k));
            }

            double scale = Math.sqrt(eigen.getEigenValues()[j]);
            for (int i = 0; i < n; i++) {
                coordinates[i][j] = eigen.getEigenVectors()[i][j] * scale;
            }
        }

        eigenvalues = eigen.getEigenValues();
        proportion = eigenvalues.clone();
        Math.unitize1(proportion);
    }
}

File: clojure/src/smile/mds.clj
;   Copyright (c) 2010-2020 Haifeng Li. All rights reserved.
;
;   Smile is free software: you can redistribute it and/or modify
;   it under the terms of the GNU Lesser General Public License as
;   published by the Free Software Foundation, either version 3 of
;   the License, or (at your option) any later version.
;
;   Smile is distributed in the hope that it will be useful,
;   but WITHOUT ANY WARRANTY; without even the implied warranty of
;   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;   GNU Lesser General Public License for more details.
;
;   You should have received a copy of the GNU Lesser General Public License
;   along with Smile.  If not, see <https://www.gnu.org/licenses/>.

(ns smile.mds
  "Multidimensional Scaling"
  {:author "Haifeng Li"}
  (:import [smile.mds MDS IsotonicMDS SammonMapping]))

(defn mds
  "Classical multidimensional scaling, also known as principal coordinates analysis.

  Given a matrix of dissimilarities (e.g. pairwise distances), MDS
  finds a set of points in low dimensional space that well-approximates the
  dissimilarities in A. We are not restricted to using a Euclidean
  distance metric. However, when Euclidean distances are used MDS is
  equivalent to PCA.

  `proximity` is the nonnegative proximity matrix of dissimilarities. The
  diagonal should be zero and all other elements should be positive and
  symmetric. For pairwise distances matrix, it should be just the plain
  distance, not squared.

  `k` is the dimension of the projection.

  If `positive` is true, estimate an appropriate constant to be added
  to all the dissimilarities, apart from the self-dissimilarities, that
  makes the learning matrix positive semi-definite. The other formulation of
  the additive constant problem is as follows. If the proximity is
  measured in an interval scale, where there is no natural origin, then there
  is not a sympathy of the dissimilarities to the distances in the Euclidean
  space used to represent the objects. In this case, we can estimate a
  constant `c` such that proximity + c may be taken as ratio data, and also
  possibly to minimize the dimensionality of the Euclidean space required for
  representing the objects."
  ([proximity k] (mds proximity k false))
  ([proximity k positive] (MDS/of proximity k positive)))

(defn isomds
  "Kruskal's nonmetric MDS.

  In non-metric MDS, only the rank order of entries in the proximity matrix
  (not the actual dissimilarities) is assumed to contain the significant
  information. Hence, the distances of the final configuration should as
  far as possible be in the same rank order as the original data. Note that
  a perfect ordinal re-scaling of the data into distances is usually not
  possible. The relationship is typically found using isotonic regression.

  `proximity` is the nonnegative proximity matrix of dissimilarities.
  The diagonal should be zero and all other elements should be positive
  and symmetric.
  `k` is the dimension of the projection.
  `tol` is the tolerance for stopping iterations.
  `max-iter` is the maximum number of iterations."
  ([proximity k] (isomds proximity k 0.0001 200))
  ([proximity k tol max-iter] (IsotonicMDS/of proximity k tol max-iter)))

(defn sammon
  "Sammon's mapping.

  The Sammon's mapping is an iterative technique for making interpoint
  distances in the low-dimensional projection as close as possible to the
  interpoint distances in the high-dimensional object. Two points close
  together in the high-dimensional space should appear close together in the
  projection, while two points far apart in the high dimensional space should
  appear far apart in the projection. The Sammon's mapping is a special case
  of metric least-square multidimensional scaling.

  Ideally when we project from a high dimensional space to a low dimensional
  space the image would be geometrically congruent to the original figure.
  This is called an isometric projection. Unfortunately it is rarely possible
  to isometrically project objects down into lower dimensional spaces. Instead
  of trying to achieve equality between corresponding inter-point distances we
  can minimize the difference between corresponding inter-point distances.
  This is one goal of the Sammon's mapping algorithm. A second goal of the
  Sammon's mapping algorithm is to preserve the topology as best as possible
  by giving greater emphasize to smaller interpoint distances. The Sammon's
  mapping algorithm has the advantage that whenever it is possible to
  isometrically project an object into a lower dimensional space it will be
  isometrically projected into the lower dimensional space. But whenever an
  object cannot be projected down isometrically the Sammon's mapping projects
  it down to reduce the distortion in interpoint distances and to limit the
  change in the topology of the object.

  The projection cannot be solved in a closed form and may be found by an
  iterative algorithm such as gradient descent suggested by Sammon. Kohonen
  also provides a heuristic that is simple and works reasonably well.

  `proximity the nonnegative proximity matrix of dissimilarities.
  The diagonal should be zero and all other elements should be positive
  and symmetric.
  `k` is the dimension of the projection.
  `lambda` is the initial value of the step size constant in diagonal Newton
  method.
  `tol` is the  tolerance for stopping iterations.
  `step-tol` is the tolerance on step size.
  `max-iter` is the maximum number of iterations."
   ([proximity k] (sammon proximity k 0.2, 0.0001 0.001 100))
   ([proximity k lambda tol step-tol max-iter] (SammonMapping/of proximity k lambda tol step-tol max-iter)))
 
